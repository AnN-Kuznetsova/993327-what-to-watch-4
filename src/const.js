const NUMBER_OF_CARDS_IN_CATALOG_AT_STARTUP = 8;
const NUMBER_OF_CARDS_TO_INCREMENT = 8;
const MAX_FILTER_ELEMENTS_COUNT = 10;
const NUMBER_OF_SIMILAR_FILMS = 4;
const VISIBLE_PARTICIPANTS_COUNT = 4;
const DELAY_PLAYBACK_PREVIEW = 1000;
const MIN_REVIEW_TEXT_LENGTH = 50;
const MAX_REVIEW_TEXT_LENGTH = 400;
const RATING_RANGE = 5;
const ERROR_COLOR = `#a8421e`;

const BASENAME_FOR_DEPLOY = `/htmlacademy-what-to-watch`;

const AppRoute = {
  MAIN: BASENAME_FOR_DEPLOY + `/`,
  SIGN_IN: BASENAME_FOR_DEPLOY + `/login`,
  MY_LIST: BASENAME_FOR_DEPLOY + `/myList`,
  FILM: BASENAME_FOR_DEPLOY + `/films/:id`,
  ADD_REVIEW: BASENAME_FOR_DEPLOY + `/films/:id/review`,
  PLAYER: BASENAME_FOR_DEPLOY + `/player/:id`,
};


export {
  DELAY_PLAYBACK_PREVIEW,
  ERROR_COLOR,
  MAX_FILTER_ELEMENTS_COUNT,
  MAX_REVIEW_TEXT_LENGTH,
  MIN_REVIEW_TEXT_LENGTH,
  NUMBER_OF_CARDS_IN_CATALOG_AT_STARTUP,
  NUMBER_OF_CARDS_TO_INCREMENT,
  NUMBER_OF_SIMILAR_FILMS,
  RATING_RANGE,
  VISIBLE_PARTICIPANTS_COUNT,
  AppRoute,
  BASENAME_FOR_DEPLOY,
};
