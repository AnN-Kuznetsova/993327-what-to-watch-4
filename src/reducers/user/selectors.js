import {NameSpace} from "../name-space.js";


const NAME_SPASE = NameSpace.USER;


const getAuthorizationStatus = (state) => {
  return state[NAME_SPASE].authorizationStatus;
};

const getLoginError = (state) => {
  return state[NAME_SPASE].loginError;
};


export {
  getAuthorizationStatus,
  getLoginError,
};
