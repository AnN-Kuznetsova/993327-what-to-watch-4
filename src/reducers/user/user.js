import {extend} from "../../utils/utils";


const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  loginError: null,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_LOGIN_ERROR: `SET_LOGIN_ERROR`,
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  setLoginError: (error) => ({
    type: ActionType.SET_LOGIN_ERROR,
    payload: error,
  }),
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password,
    })
    .then(() => {
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setLoginError(null));
    })
    .catch((error) => {
      dispatch(ActionCreator.setLoginError(error));
    });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });

    case ActionType.SET_LOGIN_ERROR:
      return extend(state, {
        loginError: action.payload
      });

    default:
      return state;
  }
};


export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
