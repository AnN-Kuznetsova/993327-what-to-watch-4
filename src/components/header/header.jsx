import PropTypes from "prop-types";
import React from "react";
import {connect} from "react-redux";

import {AuthorizationStatus} from "../../reducers/user/user";
import {ActionCreator} from "../../reducers/application/application";
import {Logo, LogoMode} from "../logo/logo";
import {PageType} from "../../const";
import {getAuthorizationStatus} from "../../reducers/user/selectors";
import {getActivePage} from "../../reducers/application/selectors";


const HeaderComponent = (props) => {
  const {
    authorizationStatus,
    activePage,
    onOpenSignInPage,
  } = props;

  const isAuth = authorizationStatus === AuthorizationStatus.AUTH;
  const isSignIn = activePage === PageType.SIGN_IN;

  const handleSignInClick = (event) => {
    event.preventDefault();
    onOpenSignInPage();
  };

  return (
    <React.Fragment>
      <h1 className="visually-hidden">WTW</h1>

      <header className={`page-header ${isAuth && !isSignIn && `movie-card__head`} ${isSignIn && `user-page__head`}`}>
        <Logo mode={LogoMode.NORMAL} />

        {isAuth && !isSignIn &&
          <div className="user-block">
            <div className="user-block__avatar">
              <img src="img/avatar.jpg" alt="User avatar" width="63" height="63" />
            </div>
          </div>}

        {!isAuth && !isSignIn &&
          <div className="user-block">
            <a href="sign-in.html" className="user-block__link"
              onClick={handleSignInClick}
            >Sign in</a>
          </div>}

        {isSignIn &&
          <h1 className="page-title user-page__title">Sign in</h1>}
      </header>
    </React.Fragment>
  );
};


HeaderComponent.propTypes = {
  authorizationStatus: PropTypes.string.isRequired,
  activePage: PropTypes.string.isRequired,
  onOpenSignInPage: PropTypes.func.isRequired,
};


const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  activePage: getActivePage(state),
});

const mapDispatchToProps = (dispatch) => ({
  onOpenSignInPage() {
    dispatch(ActionCreator.changeActivePage(PageType.SIGN_IN));
  },
});

const Header = connect(mapStateToProps, mapDispatchToProps)(HeaderComponent);

export {
  HeaderComponent,
  Header,
};
