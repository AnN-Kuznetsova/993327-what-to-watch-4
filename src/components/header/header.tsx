import * as React from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";

import {AuthorizationStatus} from "../../reducers/user/user";
import {Breadcrumbs} from "../breadcrumbs/breadcrumbs";
import {Logo, LogoMode} from "../logo/logo";
import {MovieType, PageType} from "../../types";
import {AppRoute} from "../../const";
import {getAuthorizationStatus} from "../../reducers/user/selectors";
import {getActivePage, getActiveMovie} from "../../reducers/application/selectors";


interface Props {
  authorizationStatus: string;
  activeMovie?: MovieType;
  activePage: PageType;
}


const HeaderComponent: React.FunctionComponent<Props> = (props: Props) => {
  const {
    authorizationStatus,
    activeMovie,
    activePage,
  } = props;

  const renderBreadcrumbs = activePage === PageType.ADD_REVIEW;
  const renderAvatar = authorizationStatus === AuthorizationStatus.AUTH && activePage !== PageType.ERROR && activePage !== PageType.SIGN_IN;
  const renderSignInLink = authorizationStatus === AuthorizationStatus.NO_AUTH && activePage !== PageType.ERROR && activePage !== PageType.SIGN_IN;
  const renderSignInPage = activePage === PageType.SIGN_IN;
  const renderMyList = activePage === PageType.MY_LIST;

  const breadcrambsList = activeMovie ? [
    {
      link: AppRoute.FILM.replace(`:id`, activeMovie.id),
      title: activeMovie.title
    },
    {
      link: ``,
      title: `Add review`,
    }
  ] : [];

  return (
    <React.Fragment>
      <h1 className="visually-hidden">WTW</h1>

      <header className={`page-header ${renderAvatar && !renderMyList && `movie-card__head`} ${(renderSignInPage || renderMyList) && `user-page__head`}`}>
        <Logo mode={LogoMode.NORMAL} />

        {renderBreadcrumbs &&
          <Breadcrumbs breadcrambsList={breadcrambsList} />}

        {renderAvatar &&
          <div className="user-block">
            <Link to={AppRoute.MY_LIST} >
              <div className="user-block__avatar">
                <img src="/img/avatar.jpg" alt="User avatar" width="63" height="63" />
              </div>
            </Link>
          </div>}

        {renderSignInLink &&
          <div className="user-block">
            <Link className="user-block__link"
              to={AppRoute.SIGN_IN}
            >Sign in</Link>
          </div>}

        {renderSignInPage &&
          <h1 className="page-title user-page__title">Sign in</h1>}

        {renderMyList &&
          <h1 className="page-title user-page__title">My list</h1>}
      </header>
    </React.Fragment>
  );
};


const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  activeMovie: getActiveMovie(state),
  activePage: getActivePage(state),
});

const Header = connect(mapStateToProps)(HeaderComponent);


export {
  HeaderComponent,
  Header,
};
