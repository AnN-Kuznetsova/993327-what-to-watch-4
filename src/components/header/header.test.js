import React from "react";
import renderer from "react-test-renderer";

import {AuthorizationStatus} from "../../reducers/user/user";
import {HeaderComponent} from "./header";
import {PageType} from "../../const";


global.window = Object.create(window);


describe(`Render Header`, () => {
  // TO DO: воостановить после настройки маршрутов
  /* it(`Should match with snapshot when window.location is MainPage`, () => {
    Object.defineProperty(window, `location`, {
      value: {
        pathname: `/`
      }
    });

    const headerSnapshot = renderer.create(
        <HeaderComponent mode={HeaderMode.AUTH} />
    ).toJSON();

    expect(headerSnapshot).toMatchSnapshot();
  });


  it(`Should match with snapshot when window.location is not MainPage`, () => {
    window.location.pathname = `/page-name`;

    const headerSnapshot = renderer.create(
        <HeaderComponent mode={HeaderMode.AUTH} />
    ).toJSON();

    expect(headerSnapshot).toMatchSnapshot();
  }); */


  it(`Should match with snapshot when authorizationStatus is "AUTH"`, () => {
    // window.location.pathname = `/`;
    const props = {
      authorizationStatus: AuthorizationStatus.AUTH,
      activePage: PageType.MAIN,
      onOpenSignInPage: () => {},
    };

    const headerSnapshot = renderer.create(
        <HeaderComponent {...props} />
    ).toJSON();

    expect(headerSnapshot).toMatchSnapshot();
  });


  it(`Should match with snapshot when authorizationStatus is "NO_AUTH"`, () => {
    const props = {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      activePage: PageType.MAIN,
      onOpenSignInPage: () => {},
    };

    const headerSnapshot = renderer.create(
        <HeaderComponent {...props} />
    ).toJSON();

    expect(headerSnapshot).toMatchSnapshot();
  });


  it(`Should match with snapshot when active page is "SIGN_IN"`, () => {
    const props = {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      activePage: PageType.SIGN_IN,
      onOpenSignInPage: () => {},
    };

    const headerSnapshot = renderer.create(
        <HeaderComponent {...props} />
    ).toJSON();

    expect(headerSnapshot).toMatchSnapshot();
  });


  it(`Should match with snapshot when active page is "ERROR"`, () => {
    const props = {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      activePage: PageType.ERROR,
      onOpenSignInPage: () => {},
    };

    const headerSnapshot = renderer.create(
        <HeaderComponent {...props} />
    ).toJSON();

    expect(headerSnapshot).toMatchSnapshot();
  });
});
