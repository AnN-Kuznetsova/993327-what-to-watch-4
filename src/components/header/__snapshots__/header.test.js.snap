// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Header Should match with snapshot when active page is "ERROR" 1`] = `
Array [
  <h1
    className="visually-hidden"
  >
    WTW
  </h1>,
  <header
    className="page-header false false"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link false"
        href="/"
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
  </header>,
]
`;

exports[`Render Header Should match with snapshot when active page is "SIGN_IN" 1`] = `
Array [
  <h1
    className="visually-hidden"
  >
    WTW
  </h1>,
  <header
    className="page-header false user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link false"
        href="/"
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      Sign in
    </h1>
  </header>,
]
`;

exports[`Render Header Should match with snapshot when authorizationStatus is "AUTH" 1`] = `
Array [
  <h1
    className="visually-hidden"
  >
    WTW
  </h1>,
  <header
    className="page-header movie-card__head false"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link false"
        href="/"
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="user-block"
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="img/avatar.jpg"
          width="63"
        />
      </div>
    </div>
  </header>,
]
`;

exports[`Render Header Should match with snapshot when authorizationStatus is "NO_AUTH" 1`] = `
Array [
  <h1
    className="visually-hidden"
  >
    WTW
  </h1>,
  <header
    className="page-header false false"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link false"
        href="/"
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="user-block"
    >
      <a
        className="user-block__link"
        href="sign-in.html"
        onClick={[Function]}
      >
        Sign in
      </a>
    </div>
  </header>,
]
`;
